pkg/kubectl/cmd/create_clusterrole.go

```diff
@@ -0,0 +1,286 @@
+			containsNonResourceVerbs = true
+			validVerb = true
+		}
+
+		if !validVerb {
+			return fmt.Errorf("invalid verb: '%s'", v)
+		}
+	}
+
+	if !containsNonResourceVerbs && len(c.NonResourceURLs) > 0 {
+		return fmt.Errorf("at least one non-resource verb must be specified when non-resource URL(s) provided")
+	}
+
+	if !containsResourceVerbs && len(c.Resources) > 0 {
+		return fmt.Errorf("at least one resource verb must be specified when resource(s) provided")
+	}
+
+	// validate resources.
+	mapper, _ := f.Object()
+	for _, r := range c.Resources {
+		_, err := mapper.ResourceFor(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// validate resource names, can not apply resource names to multiple resources.
+	if len(c.ResourceNames) > 0 && len(c.Resources) > 1 {
+		return fmt.Errorf("resource name(s) can not be applied to multiple resources")
+	}
+
+	// validate resource names, at least one resource must be specified if resource name(s) provided.
+	if len(c.ResourceNames) > 0 && len(c.Resources) == 0 {
+		return fmt.Errorf("at least one resource must be specified when resource name(s) provided")
+	}
+
+	if len(c.Resources) == 0 && len(c.NonResourceURLs) == 0 {
+		return fmt.Errorf("at least one resource or non-resource URL must be specified")
+	}
+
+	return nil
+}
+
+func (c *CreateClusterRoleOptions) RunCreateClusterRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
+	mapper, _ := f.Object()
+	dryRun, outputFormat := cmdutil.GetDryRunFlag(cmd), cmdutil.GetFlagString(cmd, "output")
+
+	// groupResourceMapping is a apigroup-resource map. The key of this map is api group, while the value
+	// is a string array of resources under this api group.
+	// E.g.  groupResourceMapping = {"extensions": ["replicasets", "deployments"], "batch":["jobs"]}
+	groupResourceMapping := map[string][]string{}
+
+	// This loop does the following work:
+	// 1. Constructs groupResourceMapping based on input resources.
+	// 2. Prevents pointing to non-existent resources.
+	// 3. Transfers resource short name to long name. E.g. rs.extensions is transferred to replicasets.extensions
+	for _, r := range c.Resources {
+		resource, err := mapper.ResourceFor(r)
+		if err != nil {
+			return err
+		}
+		if !arrayContains(groupResourceMapping[resource.Group], resource.Resource) {
+			groupResourceMapping[resource.Group] = append(groupResourceMapping[resource.Group], resource.Resource)
+		}
+	}
+
+	// Split verbs to resource verbs and non-resource verbs
+	resourceVerbs, nonResourceVerbs := []string{}, []string{}
+	for _, v := range c.Verbs {
+		if v == "*" {
+			resourceVerbs = []string{"*"}
+			nonResourceVerbs = []string{"*"}
+			break
+		}
+
+		if arrayContains(validResourceVerbs, v) {
+			resourceVerbs = append(resourceVerbs, v)
+		}
+
+		if arrayContains(validNonResourceVerbs, v) {
+			nonResourceVerbs = append(nonResourceVerbs, v)
+		}
+	}
+
+	clusterRole := &rbac.ClusterRole{}
+
+	// Create separate rule for each of the api group.
```

Comment:
Creating the resource policy rules should be common code between create role and create cluster role.